src/hid-asus.c

```diff
@@ -0,0 +1,320 @@
+/*
+ *  HID driver for Asus notebook built-in keyboard.
+ *  Fixes small logical maximum to match usage maximum.
+ *
+ *  Currently supported devices are:
+ *    EeeBook X205TA
+ *    VivoBook E200HA
+ *
+ *  Copyright (c) 2016 Yusuke Fujimaki <usk.fujimaki@gmail.com>
+ *
+ *  This module based on hid-ortek by
+ *  Copyright (c) 2010 Johnathon Harris <jmharris@gmail.com>
+ *  Copyright (c) 2011 Jiri Kosina
+ *
+ *  This module has been updated to add support for Asus i2c touchpad.
+ *
+ *  Copyright (c) 2016 Brendan McGrath <redmcg@redmandi.dyndns.org>
+ *  Copyright (c) 2016 Victor Vlasenko <victor.vlasenko@sysgears.com>
+ *  Copyright (c) 2016 Frederik Wenigwieser <frederik.wenigwieser@gmail.com>
+ */
+
+/*
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ */
+
+#include <linux/hid.h>
+#include <linux/module.h>
+#include <linux/input/mt.h>
+
+#include "hid-ids.h"
+
+MODULE_AUTHOR("Yusuke Fujimaki <usk.fujimaki@gmail.com>");
+MODULE_AUTHOR("Brendan McGrath <redmcg@redmandi.dyndns.org>");
+MODULE_AUTHOR("Victor Vlasenko <victor.vlasenko@sysgears.com>");
+MODULE_AUTHOR("Frederik Wenigwieser <frederik.wenigwieser@gmail.com>");
+MODULE_DESCRIPTION("Asus HID Keyboard and TouchPad");
+
+#define FEATURE_REPORT_ID 0x0d
+#define INPUT_REPORT_ID 0x5d
+
+#define INPUT_REPORT_SIZE 28
+
+#define MAX_CONTACTS 5
+
+#define MAX_X 0x0aea
+#define MAX_Y 0x06de
+#define MAX_TOUCH_MAJOR 8
+#define MAX_PRESSURE 0x80
+
+
+#define REPORT_ID_OFFSET 0
+
+#define CONTACT_DOWN_OFFSET 1
+#define CONTACT_DOWN_MASK 0x08
+
+#define BTN_LEFT_OFFSET 1
+#define BTN_LEFT_MASK 0x01
+
+#define CONTACT_DATA_OFFSET 2
+#define CONTACT_DATA_SIZE 5
+
+#define CONTACT_TOOL_TYPE_OFFSET 3
+#define CONTACT_TOOL_TYPE_MASK 0x80
+
+#define CONTACT_X_MSB_OFFSET 0
+#define CONTACT_X_MSB_BIT_SHIFT 4
+#define CONTACT_X_LSB_OFFSET 1
+
+#define CONTACT_Y_MSB_OFFSET 0
+#define CONTACT_Y_MSB_MASK 0x0f
+#define CONTACT_Y_LSB_OFFSET 2
+
+#define CONTACT_TOUCH_MAJOR_OFFSET 3
+#define CONTACT_TOUCH_MAJOR_BIT_SHIFT 4
+#define CONTACT_TOUCH_MAJOR_MASK 0x07
+
+#define CONTACT_PRESSURE_OFFSET 4
+#define CONTACT_PRESSURE_MASK 0x7f
+
+#define BYTE_BIT_SHIFT 8
+#define TRKID_SGN       ((TRKID_MAX + 1) >> 1)
+
+static void asus_report_contact_down(struct input_dev *input,
+		int toolType, u8 *data)
+{
+	int x, y, touch_major, pressure;
+
+	x = data[CONTACT_X_MSB_OFFSET] >> CONTACT_X_MSB_BIT_SHIFT;
```

Comment:
I wouldn't mind not having the macros _OFFSET, _BIT_SHIFT and just magical numbers.

Here we are decoding the protocol, so it's a little bit too much to define those numbers if they are only used once. This way, it will be easier to read the code too.

Not sure whether or not we should keep the _MASK macros however. It will depend on the general look of the code.